### PR TITLE
Add roc_auc_score metric

### PR DESCRIPTION
**Pull request by OnlyDeniko**

### CODE REVIEW

File: src/daal4py.cpp [hunk 1]
```diff
@@ -851,6 +851,46 @@ void c_train_test_split(data_or_file & orig, data_or_file & train, data_or_file
...
+
+    auto table_true = get_table(y_true);
+    auto table_test = get_table(y_test);
+
+    switch (dtype)
+    {
+        case 0:
+        {
+            float score;
+            float* predictedRank = new float[col_true];
```
Comment (PetrovKP) [inline src/daal4py.cpp:873]: Who will free this memory?

File: src/daal4py.cpp [hunk 2]
```diff
@@ -851,6 +851,46 @@ void c_train_test_split(data_or_file & orig, data_or_file & train, data_or_file
     daal::data_management::internal::trainTestSplit<int>(origTable, trainTable, testTable, trainIdxTable, testIdxTable);
 }
 
+double c_roc_auc_score(data_or_file & y_true, data_or_file & y_test, char dtype)
+{
+    const int col_true = y_true.table->getNumberOfColumns();
```
Comment (PetrovKP) [inline src/daal4py.cpp:856]:
```suggestion
    const size_t col_true = y_true.table->getNumberOfColumns();
```

File: src/daal4py.cpp [hunk 3]
```diff
@@ -851,6 +851,46 @@ void c_train_test_split(data_or_file & orig, data_or_file & train, data_or_file
...
+    const int row_test = y_test.table->getNumberOfRows();
+    if (row_true != 1 || row_test != 1 || col_true != col_test) {
+        PyErr_SetString(PyExc_RuntimeError, "Unknown shape data");
+        return NULL;
+    }
+
+    auto table_true = get_table(y_true);
+    auto table_test = get_table(y_test);
+
+    switch (dtype)
```
Comment (PetrovKP) [inline src/daal4py.cpp:868]: Use switch similar from [here](https://github.com/IntelPython/daal4py/blob/master/src/daal4py.cpp#L214)

You don't have to pass 0 and 1 and it's so awful to define the type :)

File: daal4py/sklearn/metrics/_ranking.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                                         multi_class, average, sample_weight)
+    elif y_type == "binary":
+        labels = np.unique(y_true)
+        y_true = label_binarize(y_true, classes=labels)[:, 0]
+        if len(labels) != 2:
+            raise ValueError("Only one class present in y_true. ROC AUC score "
+                         "is not defined in that case.")
+        if max_fpr is None and sample_weight is None:
+            logging.info("sklearn.metrics.roc_auc_score: " + get_patch_message("daal"))
+            if y_score.dtype == np.float64:
```
Comment (PetrovKP) [inline daal4py/sklearn/metrics/_ranking.py:40]:
```suggestion
    return d4p.daal_roc_auc_score(y_true.reshape(1, -1), y_score.reshape(1, -1))
```

File: src/daal4py.h [hunk 5]
```diff
@@ -329,6 +330,10 @@ void c_train_test_split(data_or_file & orig, data_or_file & train, data_or_file
                         data_or_file & train_idx, data_or_file & test_idx);
 }
 
+extern "C" {
+double c_roc_auc_score(data_or_file & y_true, data_or_file & y_test, char dtype);
```
Comment (PetrovKP) [inline src/daal4py.h:334]:
```suggestion
double c_roc_auc_score(data_or_file & y_true, data_or_file & y_test);
```

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/eb2dc918-5433-f1e0-9437-8cdcd4b723a1?

File: src/daal4py.cpp [hunk 6]
```diff
@@ -851,6 +851,45 @@ void c_train_test_split(data_or_file & orig, data_or_file & train, data_or_file
...
+    auto table_true = get_table(y_true);
+    auto table_test = get_table(y_test);
+    auto type = (*table_test->getDictionary())[0].indexType;
+    if (type == daal::data_management::data_feature_utils::DAAL_FLOAT64 ||
+        type == daal::data_management::data_feature_utils::DAAL_INT64_S ||
+        type == daal::data_management::data_feature_utils::DAAL_INT64_U)
+    {
+        double score;
+        double* predictedRank = new double[col_true];
+        daal::data_management::internal::calculateRankData<double>(predictedRank, table_test, col_true);
```
Comment (PetrovKP) [inline src/daal4py.cpp:874]: Merge to one function for example

```suggestion
        score = daal::data_management::internal::rocAucScore<double>(table_true, table_test);
```

File: daal4py/sklearn/monkeypatch/dispatcher.py [hunk 7]
```diff
@@ -91,6 +93,7 @@ def _get_map_of_algorithms():
         'knn_regressor':            [[(neighbors_module, 'KNeighborsRegressor', KNeighborsRegressor_daal4py), None]],
         'random_forest_classifier': [[(ensemble_module, 'RandomForestClassifier', RandomForestClassifier_daal4py), None]],
         'random_forest_regressor':  [[(ensemble_module, 'RandomForestRegressor', RandomForestRegressor_daal4py), None]],
+        'roc_auc_score':            [[(metrics, 'roc_auc_score', _daal_roc_auc_score), None]],
```
Comment (PetrovKP) [inline daal4py/sklearn/monkeypatch/dispatcher.py:96]: You need to add a check for version 2021.2 - so that CI will work.

File: daal4py/sklearn/metrics/_ranking.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        if multi_class == 'raise':
+            raise ValueError("multi_class must be in ('ovo', 'ovr')")
+        logging.info("sklearn.metrics.roc_auc_score: " + get_patch_message("sklearn"))
+        return _multiclass_roc_auc_score(y_true, y_score, labels,
+                                         multi_class, average, sample_weight)
+    elif y_type == "binary":
+        labels = np.unique(y_true)
+        y_true = label_binarize(y_true, classes=labels)[:, 0]
+        if max_fpr is None and sample_weight is None and len(labels) == 2:
+            logging.info("sklearn.metrics.roc_auc_score: " + get_patch_message("daal"))
```
Comment (Pahandrovich) [inline daal4py/sklearn/metrics/_ranking.py:53]: I think we need to add additional [check](https://github.com/scikit-learn/scikit-learn/blob/0fb307bf39bbdacd6ed713c00724f8f871d60370/sklearn/metrics/_ranking.py#L230) here.

**Comment by Pahandrovich** [comment]
http://intel-ci.intel.com/eb307d7d-75cf-f167-8e34-8cdcd4b723a1

**Review by Pahandrovich — APPROVED** [review]
LGTM

**Review by PetrovKP — APPROVED** [review]
